x-pack/plugin/ccr/src/internalClusterTest/java/org/elasticsearch/xpack/ccr/CcrRepositoryIT.java

```diff
@@ -613,83 +602,76 @@ public void testCcrRepositoryFailsToFetchSnapshotShardSizes() throws Exception {
+                        simulatedFailures.incrementAndGet();
                         channel.sendResponse(new ElasticsearchException("simulated"));
-                        return;
                     }
                 }
                 handler.messageReceived(request, channel, task);
             });
-            transportServices.add(mockTransportService);
         }
 
         final String followerIndex = "follower";
         try {
-            final String leaderCluster = CcrRepository.NAME_PREFIX + "leader_cluster";
-            final RepositoriesService repositoriesService = getFollowerCluster().getCurrentMasterNodeInstance(RepositoriesService.class);
-            final Repository repository = repositoriesService.repository(leaderCluster);
-            assertThat(repository.getMetadata().type(), equalTo(CcrRepository.TYPE));
-            assertThat(repository.getMetadata().name(), equalTo(leaderCluster));
-
-            for (int i = 0; i < numberOfShards; i++) {
-                final Index index = indexMetadata.getIndex();
-                final int shardId = i;
-                ElasticsearchException exception = expectThrows(ElasticsearchException.class,
-                    () -> repository.getShardSnapshotStatus(
-                        new SnapshotId(CcrRepository.LATEST, CcrRepository.LATEST),
-                        new IndexId(index.getName(), index.getUUID()),
-                        new ShardId(index, shardId)));
-                assertThat(exception.getMessage(), equalTo("simulated"));
-            }
-            assertThat(indicesStatsRequestsCount.getAndSet(0), equalTo(numberOfShards));
-
-            final RestoreService restoreService = getFollowerCluster().getCurrentMasterNodeInstance(RestoreService.class);
-            final ClusterService clusterService = getFollowerCluster().getCurrentMasterNodeInstance(ClusterService.class);
+            final SnapshotsInfoService snapshotsInfoService = getFollowerCluster().getCurrentMasterNodeInstance(SnapshotsInfoService.class);
 
-            final PlainActionFuture<IndexRoutingTable> waitForRestoreInProgress = PlainActionFuture.newFuture();
+            final PlainActionFuture<List<Long>> waitForAllShardSnapshotSizesFailures = PlainActionFuture.newFuture();
             final ClusterStateListener listener = event -> {
                 RestoreInProgress restoreInProgress = event.state().custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY);
                 if (restoreInProgress != null
                     && restoreInProgress.isEmpty() == false
                     && event.state().routingTable().hasIndex(followerIndex)) {
-                    waitForRestoreInProgress.onResponse(event.state().routingTable().index(followerIndex));
+                    try {
+                        final IndexRoutingTable indexRoutingTable = event.state().routingTable().index(followerIndex);
+                        assertBusy(() -> {
```

Comment:
It seems a bit delicate to wait inside the cluster-listener, I think that could lead to strange failures. On the other hand, it seems quite deliberate, since the old version did the assertBusy outside the listener. I would be curious to know more details on why this needed to be like this?

Reply:
`InternalSnapshotsInfoService` listens for cluster state updates while shard snapshot size fetchings are executing in another thread pool (triggering more cluster state updates when the fetching finishes). With batched cluster states and intermediate reroutes, it is hard to catch the moment where all shards are still unassigned with `FETCHING_SHARD_DATA` while the `InternalSnapshotsInfoService` has the result of the size fetchings... 

Because size fetching results are cleared up once corresponding shards are assigned, using `assertBusy()` here is the only way I can think of to catch the exact values fetched by the service.

The previous test was inadapted as shards remain unassigned until a reroute was triggered, whereas we want `InternalSnapshotsInfoService` to move the allocation of shards forward even in case of failure.

Reply:
Can we add a comment on the importance of the ordering of cluster service listeners? Something like:

// this assertBusy completes because the listener is added after the `InternalSnapshotsInfoService` and `ClusterService` preserves the order of listeners.

I think that such a comment can help if this breaks in the future...

Reply:
Sure, thanks, I'll borrow your comment.